libcnb/src/layer/struct_api/mod.rs

```diff
@@ -0,0 +1,228 @@
+    /// The layer was cached in a previous buildpack run, but the metadata was invalid and couldn't
+    /// be converted into a valid form. Subsequently, the layer was deleted entirely.
+    ///
+    /// See: `invalid_metadata` in [`CachedLayerDefinition`].
+    MetadataInvalid { cause: MAC },
+    /// The layer was cached in a previous buildpack run, but the `inspect_restored` function
+    /// rejected the contents and/or metadata.
+    ///
+    /// See: `inspect_restored` in [`CachedLayerDefinition`].
+    Inspect { cause: RAC },
+}
+
+/// A value-to-value conversion for layer actions.
+///
+/// Similar to [`Into`], but specialized. Allowing it to also be implemented for
+/// values in the standard library such as [`Result`].
+///
+/// Implement this trait if you want to use your own types as actions.
+///
+/// libcnb ships with generic implementations for the majority of the use-cases:
+/// - Using [`InspectRestoredAction`] or [`InvalidMetadataAction`] directly.
+/// - Using [`InspectRestoredAction`] or [`InvalidMetadataAction`] directly, wrapped in a Result.
+/// - Using [`InspectRestoredAction`] or [`InvalidMetadataAction`] with a cause value in a tuple.
+/// - Using [`InspectRestoredAction`] or [`InvalidMetadataAction`] with a cause value in a tuple, wrapped in a Result.
+pub trait IntoAction<T, C, E> {
+    fn into_action(self) -> Result<(T, C), E>;
+}
+
+// Allows to use the layer actions directly.
+impl<T, E> IntoAction<T, (), E> for T {
+    fn into_action(self) -> Result<(T, ()), E> {
+        Ok((self, ()))
+    }
+}
+
+//  Allows to use the layer actions directly wrapped in a Result.
+impl<T, E> IntoAction<T, (), E> for Result<T, E> {
+    fn into_action(self) -> Result<(T, ()), E> {
+        self.map(|value| (value, ()))
+    }
+}
+
+// Allows to use the layer actions directly with a cause as a tuple.
+impl<T, C, E> IntoAction<T, C, E> for (T, C) {
+    fn into_action(self) -> Result<(T, C), E> {
+        Ok(self)
+    }
+}
+
+// Allows to use the layer actions directly with a cause as a tuple wrapped in a Result.
+impl<T, C, E> IntoAction<T, C, E> for Result<(T, C), E> {
+    fn into_action(self) -> Result<(T, C), E> {
+        self
+    }
+}
+
+/// A reference to an existing layer on disk.
+///
+/// Provides functions to modify the layer such as replacing its metadata, environment, SBOMs or
+/// exec.d programs.
+///
+/// To obtain a such a reference, use [`BuildContext::cached_layer`] or [`BuildContext::uncached_layer`].
+pub struct LayerRef<B, MAC, RAC>
+where
+    B: Buildpack + ?Sized,
+{
+    name: LayerName,
+    // Technically not part of the layer itself. However, the functions that modify the layer
+    // will need a reference to the layers directory as they will also modify files outside the
+    // actual layer directory. To make LayerRef nice to use, we bite the bullet and include
+    // the layers_dir here.
+    layers_dir: PathBuf,
+    buildpack: PhantomData<B>,
+    pub state: LayerState<MAC, RAC>,
+}
```

Comment:
Suggestion: Add a `fn read_env` to LayerRef.

Context:
With the current layer design I have a pattern of applying the env of the newly created layer to future layers https://github.com/heroku/buildpacks-ruby/blob/aaa72f317ee7b4700de35ea27112c1ad15334def/buildpacks/ruby/src/main.rs#L146. This is important to do even if the buildpack did not write any `LayerEnv` data because the `bin/` path of the layer is always put on the path.

Therefore, I'm wanting some mechanism of pulling out the current `LayerEnv` of a layer. This exists:

```
  LayerEnv::read_from_layer_dir(barnes_layer.path())
      .expect("read env from a layer")
      .apply(Scope::Build, &env),
```

However, the docs say not to use it, and it returns `std::io::Error` instead of a `libcnb::Error`. 


Sorry for tacking this somewhere random, but it seemed to make the most sense to put the request at the end of `LayerRef`.

Reply:
I was just about to ask the same question :-) I'm currently using `LayerEnv::read_from_layer_dir(layer_ref.path()`, which works great, but the warning not to use it should maybe be changed (if not an alternative implementation that works with the new API.